sycl/include/sycl/ext/oneapi/properties/properties.hpp

```diff
@@ -9,265 +9,180 @@
+template <class T>
+struct is_empty_or_incomplete<
+    T, std::enable_if_t<(sizeof(T) > 0) && !std::is_empty_v<T>>>
+    : std::false_type {};
+template <class T>
+constexpr bool is_empty_or_incomplete_v = is_empty_or_incomplete<T>::value;
+template <class T>
+using IsRuntimeProperty = mp11::mp_not<is_empty_or_incomplete<T>>;
 } // namespace detail
 
-template <typename PropertiesT> class properties {
-  static_assert(detail::IsTuple<PropertiesT>::value,
-                "Properties must be in a tuple.");
-  static_assert(detail::AllPropertyValues<PropertiesT>::value,
+// base case is only for invalid cases to suppress compiler errors about base
+// class storage of properties
+template <class L, class = void> class properties {
+  static_assert(detail::mp11::mp_all_of<L, std::is_class>(),
                 "Unrecognized property in property list.");
-  static_assert(detail::IsSorted<PropertiesT>::value,
-                "Properties in property list are not sorted.");
-  static_assert(detail::SortedAllUnique<PropertiesT>::value,
+  static_assert(detail::mp11::mp_is_set<L>(),
                 "Duplicate properties in property list.");
-  static_assert(detail::NoConflictingProperties<PropertiesT>::value,
-                "Conflicting properties in property list.");
-
-public:
-  template <typename... PropertyValueTs>
-  constexpr properties(PropertyValueTs... props)
-      : Storage(detail::ExtractProperties<StorageT>::Extract(
-            std::tuple<PropertyValueTs...>{props...})) {}
+};
 
-  template <typename PropertyT>
-  static constexpr std::enable_if_t<detail::IsProperty<PropertyT>::value, bool>
-  has_property() {
-    return detail::ContainsProperty<PropertyT, PropertiesT>::value;
-  }
+template <template <class...> class TL, class... V>
+class properties<TL<V...>,
+                 std::enable_if_t<(std::is_class_v<V> && ...) &&
+                                  detail::mp11::mp_is_set<TL<V...>>::value>>
+    : V... {
+private:
+  using map = detail::property_map<properties>; // map of properties
+  static_assert(detail::mp11::mp_is_map<map>(),
+                "Duplicate properties in property list.");
+  using keys = detail::mp11::mp_map_keys<map>;
+  using conflicting = detail::mp11::mp_apply<
+      detail::mp11::mp_append,
+      detail::mp11::mp_transform<detail::ConflictingProperties, keys>>;
+  static_assert(
+      !detail::mp11::mp_any_of_q<
+          conflicting,
+          detail::mp11::mp_bind_front<detail::mp11::mp_set_contains, keys>>(),
+      "Conflicting properties in property list.");
 
-  template <typename PropertyT>
-  typename std::enable_if_t<detail::IsRuntimeProperty<PropertyT>::value &&
-                                has_property<PropertyT>(),
-                            PropertyT>
-  get_property() const {
-    return std::get<PropertyT>(Storage);
+public:
+  template <class... T>
+  constexpr properties(T... v)
+      : T(v)... {} // T might have different ordering than V
```

Comment:
Nit; Maybe personal preference, but I would like these to have a single space between each function/ctor definition. It too me way too long to find this ctor because it looked like it was associated with the `has_property` below.

Also, does the comment matter?

Reply:
the comment would probably be better on either of the previous lines. I wanted to explain why it needs to be as-is and cannot be simplified from ` template <class... T> constexpr properties(T... v)` to  `constexpr properties(V... v)`.